parser/src/string.rs

```diff
@@ -0,0 +1,155 @@
+use crate::ast::{Constant, Expr, ExprKind, Location};
+use crate::error::LexicalErrorType::FStringError;
+use crate::error::{LexicalError, LexicalErrorType};
+use crate::fstring::parse_located_fstring;
+use crate::token::StringKind;
+use itertools::Itertools;
+
+pub fn parse_implicit_concatenation(
+    values: Vec<(Location, (String, StringKind))>,
+) -> Result<Expr, LexicalError> {
+    // Preserve the initial location and kind.
+    let initial_location = values[0].0;
+    let initial_kind = (values[0].1.1 == StringKind::U).then(|| "u".to_owned());
+
+    // Determine whether the list of values contains any f-strings. (If not, we can return a
+    // single Constant at the end, rather than a JoinedStr.)
+    let has_fstring = values
+        .iter()
+        .any(|(_, (_, string_kind))| *string_kind == StringKind::F);
+
+    // De-duplicate adjacent constants.
+    let mut deduped: Vec<Expr> = vec![];
+    let mut current: Vec<String> = vec![];
+    for (location, (string, string_kind)) in values {
+        match string_kind {
+            StringKind::Normal => current.push(string),
+            StringKind::U => current.push(string),
+            StringKind::F => {
+                if let ExprKind::JoinedStr { values } = parse_located_fstring(&string, location)
```

Comment:
Arguably `parse_located_fstring` could now return `Vec<Expr>`?

Reply:
oh, that makes sense! we never use the returned JoinedStr

Reply:
I can do this in a separate PR.